stdlib/builtins.pyi

```diff
@@ -1477,16 +1477,23 @@ def sorted(
 ) -> list[SupportsRichComparisonT]: ...
 @overload
 def sorted(__iterable: Iterable[_T], *, key: Callable[[_T], SupportsRichComparison], reverse: bool = ...) -> list[_T]: ...
+
+class _SupportsSum(Protocol):
+    def __add__(self, __x: Any) -> Any: ...
+
+_SumT = TypeVar("_SumT", bound=_SupportsSum)
+_SumS = TypeVar("_SumS", bound=_SupportsSum)
+
 @overload
-def sum(__iterable: Iterable[_T]) -> _T | Literal[0]: ...
+def sum(__iterable: Iterable[_SumT]) -> Any | Literal[0]: ...
```

Comment:
In 90% of cases, I think people will probably be using `sum` to add up variables that are all of the same type, e.g. `sum((1.4, 2.3, 3.7, 4.2))`. We can safely say for this that the result will be of type `float`.

So, although strictly speaking you're entirely correct — the return type isn't guaranteed to always be `_SumT`, far from it! — I nonetheless think I prefer using `_SumT` over `Any` for the return type, so that we can preserve precise type inference for the common case.

But I'm curious about what other people think :)

Reply:
Also, note that the TypeVar is currently unbound in this overload, so either the return type needs to use `_SumT` instead of `Any`, or the parameter type needs to change to use `_SupportsSum` instead of `_SumT` :)

Reply:
What if we make the protocol generic over the return type of `__add__`, and make it `sum(Iterable[_SupportsSum[_T]]) -> _T`?

Reply:
Ahh shoot this was actually unintentional, was playing round with the typings and forgot to change it back 😅 
I agree that `_SumT` makes sense, so will change it back for now 👍

Reply:
> What if we make the protocol generic over the return type of `__add__`, and make it `sum(Iterable[_SupportsSum[_T]]) -> _T`?

Good shout, will give it a go

Reply:
@JelleZijlstra, not sure it's that simple, because there's `__radd__` to consider, which can change the return type. If Y is a subclass of X, and Y defines `__radd__`, then `Y.__radd__ `takes precedence over `X.__add__` when evaluating the operation `x + y`

Reply:
Idk  how much that actually matters though

Reply:
Yes, getting a completely precise return type in the face of mixed types and `__radd__` is very complicated. The current version is technically incorrect for the case where two arguments are passed, the iterable is empty, and the default returns an unrelated type from its `__add__`.

mypy-primer found a lot of new false positives from my suggestion, so we should probably undo that and just go back to assuming that `__add__` returns the same type.

Reply:
Yeah I've reverted back to using TypeVars in the latest commit 😄